src/helpers/unstable_mapSliceZone.ts

```diff
@@ -0,0 +1,266 @@
+	TContext = unknown,
+> = (args: MapperArgs<TSlice, TContext>) => TProps | Promise<TProps>;
+
+/**
+ * Arguments provided to a mapper function.
+ */
+export type MapperArgs<
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	TSlice extends SliceLike = any,
+	TContext = unknown,
+> = SliceComponentProps<TSlice, TContext>;
+
+/**
+ * Unwraps a lazily loaded mapper module.
+ */
+type ResolveLazyMapperModule<TMapper extends Mapper | LazyModule<Mapper>> =
+	TMapper extends LazyModule<Mapper>
+		? Awaited<ReturnType<TMapper>> extends {
+				// eslint-disable-next-line @typescript-eslint/no-explicit-any
+				default: any;
+		  }
+			? Awaited<ReturnType<TMapper>>["default"]
+			: Awaited<ReturnType<TMapper>>
+		: TMapper;
+
+/**
+ * Transforms a Slice into its mapped version.
+ */
+type MapSliceLike<
+	TSliceLike extends SliceLike,
+	TMappers extends Record<string, Mapper>,
+> = TSliceLike extends SliceLikeRestV2
+	? TSliceLike["slice_type"] extends keyof TMappers
+		? SliceLikeRestV2<TSliceLike["slice_type"]> &
+				MappedSliceLike &
+				Awaited<
+					ReturnType<
+						ResolveLazyMapperModule<TMappers[TSliceLike["slice_type"]]>
+					>
+				>
+		: TSliceLike
+	: TSliceLike extends SliceLikeGraphQL
+	? TSliceLike["type"] extends keyof TMappers
+		? SliceLikeGraphQL<TSliceLike["type"]> &
+				MappedSliceLike &
+				Awaited<
+					ReturnType<ResolveLazyMapperModule<TMappers[TSliceLike["type"]]>>
+				>
+		: TSliceLike
+	: never;
+
+/**
+ * Transforms a Slice Zone using a set of mapping functions, one for each type
+ * of Slice. Mapping functions can be async.
+ *
+ * Whenever possible, use this function on the server to minimize client-side
+ * processing.
+ *
+ * @example
+ *
+ * ```typescript
+ * const mappedSliceZone = await unstable_mapSliceZone(page.data.slices, {
+ * 	code_block: ({ slice }) => ({
+ * 		codeHTML: await highlight(slice.primary.code),
+ * 	}),
+ * });
+ * ```
+ *
+ * @experimental Names and implementations may change in the future.
+ * `unstable_mapSliceZone()` does not follow SemVer.
+ */
+export function unstable_mapSliceZone<
+	TSliceLike extends SliceLike,
+	TMappers extends Record<string, Mapper>,
+	TContext = unknown,
+>(
+	sliceZone: SliceZoneLike<TSliceLike>,
+	mappers: TMappers,
+	context?: TContext,
+): Promise<MapSliceLike<TSliceLike, TMappers>[]> {
+	return Promise.all(
+		sliceZone.map(async (slice, index, slices) => {
+			const isRestSliceType = "slice_type" in slice;
+			const sliceType = isRestSliceType ? slice.slice_type : slice.type;
+
+			const mapper = mappers[sliceType];
+
+			if (!mapper) {
+				return slice;
+			}
+
+			const mapperArgs = { slice, slices, index, context };
+
+			let result = await mapper(mapperArgs);
+
+			if (
+				mapper.length < 1 &&
+				(typeof result === "function" ||
+					(typeof result === "object" && "default" in result))
+			) {
+				result = "default" in result ? result.default : result;
+				result = await result(mapperArgs);
```

Comment:
I'd maybe rename the first `result` to something like `resultOrMapperModule` so that the following line can read better.

Reply:
I prefer `result` here since we ultimately will be using `result` as the final mapper function return value. `result` is used later in the function to spread into an object, which should never be a module.

However, I added comments around `const result` to explain exactly what is happening at that stage. I think that should accomplish what your rename suggestion is doing.

What do you think?

Reply:
Works for me!